List and Manage list components display the list name and who owns the list

rachelspencer opened the pull request:
## Description

Users can now see the name of the list when they're on the `List` and `Manage List` page. They can also see who owns the list on the `List` and `Manage List` page. 

## Related Issue

closes #18

## Acceptance Criteria

- [x] The `List` component renders copy that displays the list name.
- [x] The `Manage List` component renders copy that displays the list name.
- [x] The `List` component renders copy that displays who owns the list.
- [x] `Manage List` component renders copy that displays who owns the list.

## Type of Changes

Enhances

## Updates

### Before
List 
<img width="1649" alt="List before" src="https://github.com/the-collab-lab/tcl-73-smart-shopping-list/assets/111473039/f414809c-88e4-4325-ab01-c1cdda427876">

Manage List
<img width="1510" alt="Manage List before" src="https://github.com/the-collab-lab/tcl-73-smart-shopping-list/assets/111473039/dd9e3f29-7e8f-4853-b29b-0271172738b3">

### After
Your list with items
<img width="1700" alt="Your list with items" src="https://github.com/the-collab-lab/tcl-73-smart-shopping-list/assets/111473039/bfe16143-a6f3-4d83-8df3-727bc4a47efd">

Your list with no items
<img width="1232" alt="Your list no items" src="https://github.com/the-collab-lab/tcl-73-smart-shopping-list/assets/111473039/80d9fa4f-5c6d-4ee4-8d4b-f395f6921d00">

Shared list with items
<img width="1137" alt="Screenshot 2024-03-27 at 4 16 24 PM" src="https://github.com/the-collab-lab/tcl-73-smart-shopping-list/assets/111473039/92b02031-94ee-4eec-b90a-02940c74c4bb">

Shared list with no items
<img width="979" alt="Screenshot 2024-03-27 at 4 17 00 PM" src="https://github.com/the-collab-lab/tcl-73-smart-shopping-list/assets/111473039/3efe8817-d5e2-401f-920a-7adca62a97bc">


Your list on Manage List
<img width="981" alt="Manage List yours" src="https://github.com/the-collab-lab/tcl-73-smart-shopping-list/assets/111473039/f6fce492-e321-4700-b211-48305aaa73e1">

Shared list on Manage List
<img width="1019" alt="Screenshot 2024-03-27 at 4 17 19 PM" src="https://github.com/the-collab-lab/tcl-73-smart-shopping-list/assets/111473039/af8f8e7d-466a-4f2e-abaf-81e9cf938e35">

## Testing Steps / QA Criteria

- git pull
- git checkout rs-display-list-name
- npm ci
- npm start

Checking list name
- click on any list in your `Home` component
- navigate to `List` and `Manage List` and check that the selected list name appears at the top

Checking list ownership messages
Your list
- navigate to the `Home` component
- create a new list name 
- navigate to `List` and `Manage List` and check that there is a message at that says  `You own this list`

A Shared List
- navigate to the `Home` component
- Select the list called `CB2` (this is a new list shared with you by me)
- navigate to `List` and `Manage List` and check that there is a message at that says  `This list belongs to Rachel Spencer`
- This list should have one item on the `List` component

- navigate to the `Home` component
- Select the list called `Coffee Supply` (this is a new list shared with you by me)
- navigate to `List` and `Manage List` and check that there is a message at that says  `This list belongs to Rachel Spencer`
- This list should have no items on the `List` component

Important note - all existing shared lists created before this issue won't display the name of who owns the shared list. It's recommended that all users clean up their FireStore data and remove all old shared lists.

rachelspencer committed:
local storage setting list name, list name console logging on List Component

rachelspencer committed:
p tag rendering list name for a list with no items

rachelspencer committed:
p tag rendering list name for a list with items

rachelspencer committed:
pass listName prop to ManageList component from App.js and render list name with a p tag

rachelspencer committed:
found bug in solution so removed logic to start fresh

rachelspencer committed:
fix typo in shareList error message

rachelspencer committed:
begin to create helper hook that gets list name from path

rachelspencer committed:
complete helper hook to get list name from list path and removed console log from Home componenet

rachelspencer committed:
added listName prop to ManageList, rendered list name on ManageList page

rachelspencer committed:
added list name for lists with no items

rachelspencer committed:
begin creating helper hook to identify owner of list from path

rachelspencer committed:
p tag rendering at the bottom of the List component displaying the list owner status

rachelspencer committed:
firebase function created to get list owner name, start to render this detail on the List component

rachelspencer committed:
successfully render the name of a shared list on the List component

rachelspencer committed:
refactor list owner jxs into eliminate duplication of code

rachelspencer committed:
create new component that will render owner message on List and ManageList page

rachelspencer committed:
remove console.log

rachelspencer committed:
create custom component that handles rendering who the list is owned by, implemented in the List component

rachelspencer committed:
custom hook made to get current list owner details, imlement hook in List component

rachelspencer committed:
render owner list name details in Manage List component

rachelspencer committed:
refactor so useListOwnerDetails is only called from ListOwnerMessaage instead of in List and ManageList

rachelspencer committed:
remove console logs in ListOwnerMessage

rachelspencer committed:
clean up getListOwnerDetails function in firebase

trushmi reviewed (approved):
Hi Raichel! Great work tackling this PR! There are a couple of edge cases that need improvement related to the scenario where the user is logged out or not signed in and tries to access the application.

1. When the user is signed in, everything works as expected. However, when the user logs out and opens the project in an anonymous window, the website doesn’t function properly, and there is nothing on the page. Error: Cannot read properties of null (reading 'split')

```jsx
javascriptCopy code
const indexOfFwdSlash = listPath.split('').indexOf('/');

```

2. In case the user signs out and still has the browser open, instead of the owner's name, there is an empty space:

"This list belongs to ."

Please, take a look on these issues.

rachelspencer committed:
add logic to conditionally render list name and ownership if a user is logged in

rachelspencer commented:
Hi @trushmi, 

Thanks for your comments! I have added logic to only render the list ownership message if a user is logged in. This handles both of the issues you spotted.

This made me also realize that the list name should only be shown if a user is logged in, so I added logic to handle this as well.

jaguilar89 reviewed (approved):
Hi Rachel! Code looks good, and everything works as expected in accordance with the testing criteria. I too noticed the bug where the list name, owner name, and list items were still being displayed after signing out of the app, but looks like you took care of it. Good job!!

sillytsundere reviewed (approved):
Everything looks great to me Rachel, and the code tested as desired! Great job! This became a lot more than we thought it was going to be so really great work! :3

timothy-taylor commented on src/utils/hooks.js:
```diff
@@ -19,3 +20,26 @@ export function useStateWithStorage(storageKey, initialValue) {
 	return [value, setValue];
 }
+
+export function useListOwnerDetails(userIdFromPath, listName) {
```
Well done with this hook!

rachelspencer replied:
Thank you!

timothy-taylor commented on src/utils/listPath.js:
```diff
@@ -0,0 +1,14 @@
+export const getListNameFromPath = (listPath) => {
+	const indexOfFwdSlash = listPath.split('').indexOf('/');
```
when I open the live build of the PR branch I am getting a `Cannot read properties of null (reading 'split')` error and nothing is loading. I would assume its happening due to the `.split` call in both of these functions being called in App.jsx. Maybe add some defensive programming to these so they don't error on initial render? Either `listPath?.split....` or a `if (!listPath) return` early return?

rachelspencer replied:
Thanks Tim, I also added an extra conditional to only render the list name if the user is signed in and there is a list name. I saw that the list name message was appearing with no actual name when I ran through the app in incognito mode once more, after signing in for the first time and navigating to `List` and `Manage List` before selecting a list name.

![Screenshot 2024-03-29 at 10 14 56 AM](https://github.com/the-collab-lab/tcl-73-smart-shopping-list/assets/111473039/8ca63eb2-f7bf-4fce-ad0f-1cebc038797c)

timothy-taylor reviewed (approved):
Great work on this! The code is clean and well structured. I left a comment in regards to adding some defensive programming to your listPath util functions, but I tested it locally and everything looks good!

rachelspencer committed:
add early return to getListNameFromPath and getUserIdFromPath hooks

rachelspencer committed:
add conditional to only render the list name if there is a current user and a list name
